### ISSUE TITLE
[Bug]: Plex Libraries Not Populating

### ISSUE THREAD

**Issue by kagrith** [chrisbenincasa/tunarr#602]
### What happened?

Attempting to add libraries, and tunarr does not populate, but server shows as valid connection.
Add media page shows no libraries or shows.
![image](https://github.com/chrisbenincasa/tunarr/assets/9443416/9cc1b2da-67bf-4735-a1bc-aa75130d7ae3)
![image](https://github.com/chrisbenincasa/tunarr/assets/9443416/f8963c06-51b3-43fe-ba71-96279f4a31ee)
![image](https://github.com/chrisbenincasa/tunarr/assets/9443416/80041a36-d0e6-4971-9628-55623c903938)

Logs sanitized.

### Version

0.1.0

### What browser did you experience the the problem on?

Firefox 126.0.1

### What operating system are you using?

Win 10

### Relevant log output

```shell
_____ _   _ _  _   _   ___ ___ 
|_   _| | | | \| | /_\ | _ \ _ \
  | | | |_| | .` |/ _ \|   /   /
  |_|  \___/|_|\_/_/ \_\_|_\_|_\

                0.1.0

2024-07-02T18:02:35.586Z [info]: Searching for legacy dizquetv directory at /.dizquetv
/tunarr/server/build/.tunarr/stream-cache.json null
2024-07-02T18:02:35.835Z [info]: XMLTV Updated at 7/2/2024, 1:02:35 PM {"task":"update-xmltv"}
2024-07-02T18:02:35.855Z [info]: HTTP server running on port: http://0.0.0.0:8000
2024-07-02T18:03:02.012Z [info]: Open event channel.
2024-07-02T18:03:06.677Z [error]: Plex request error: getaddrinfo ENOTFOUND myPlexServerIP.myPlexDirectID.plex.direct 
    err: {
      "message": "getaddrinfo ENOTFOUND myPlexServerIP.myPlexDirectID.plex.direct: getaddrinfo ENOTFOUND myPlexServerIP.myPlexDirectID.plex.direct",
      "name": "Error",
      "stack":
          Error: getaddrinfo ENOTFOUND myPlexServerIP.myPlexDirectID.plex.direct
              at AxiosError.from (file:///tunarr/server/build/bundle.js:241576:14)
              at RedirectableRequest.handleRequestError (file:///tunarr/server/build/bundle.js:243323:34)
              at RedirectableRequest.emit (node:events:519:28)
              at eventHandlers.<computed> (file:///tunarr/server/build/bundle.js:121942:28)
              at ClientRequest.emit (node:events:519:28)
              at TLSSocket.socketErrorListener (node:_http_client:500:9)
              at TLSSocket.emit (node:events:519:28)
              at emitErrorNT (node:internal/streams/destroy:169:8)
              at emitErrorCloseNT (node:internal/streams/destroy:128:3)
              at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
          caused by: Error: getaddrinfo ENOTFOUND myPlexServerIP.myPlexDirectID.plex.direct
              at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:120:26)
              at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17)
      "config": {
        "transitional": {
          "silentJSONParsing": true,
          "forcedJSONParsing": true,
          "clarifyTimeoutError": false
        },
        "adapter": [
          "xhr",
          "http"
        ],
        "transformRequest": [
          null
        ],
        "transformResponse": [
          null
        ],
        "timeout": 0,
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
        "maxContentLength": -1,
        "maxBodyLength": -1,
        "env": {},
        "headers": {
          "Accept": "application/json",
          "X-Plex-Device": "Tunarr",
          "X-Plex-Device-Name": "Tunarr",
          "X-Plex-Product": "Tunarr",
          "X-Plex-Version": "0.1",
          "X-Plex-Client-Identifier": "p86cy1w47clco3ro8t92nfy1",
          "X-Plex-Platform": "Chrome",
          "X-Plex-Platform-Version": "80.0",
          "X-Plex-Token": "myPlexToken",
          "User-Agent": "axios/1.6.0",
          "Accept-Encoding": "gzip, compress, deflate, br"
        },
        "baseURL": "https://myPlexServerIP.myPlexDirectID.plex.direct:32400",
        "method": "get",
        "url": "/library/sections",
        "metadata": {
          "startTime": 1719943386436
        }
      },
      "code": "ENOTFOUND",
      "status": null
    }
2024-07-02T18:03:41.459Z [info]: Open event channel.
```

**Comment by chrisbenincasa** [comment]
Hello, sorry for the delay, I was on vacation.

It looks like the Tunarr server is unable to reach your Plex server over the network, for some reason. It's definitely strange that the backend status is reporting healthy, however. It could be that for the status check, the DNS was able to resolve (for some reason). How are you running Tunarr?

**Comment by chrisbenincasa** [comment]
Thanks. And I'm guessing that on your Plex settings page, the backend connection to Plex doesn't show an error?

**Comment by kagrith** [comment]
Correct. Everything appears nominal.

When I switch to direct playback nothing changes either.

Could we retain the ability to swap to a local IP address to simplify
routing?

In my opinion, the latency and complexity of going over plex direct when
the local network is available is a waste of bandwidth.

**Comment by chrisbenincasa** [comment]
You can directly edit the Plex url right in the settings page by clicking on the pencil icon

**Comment by kagrith** [comment]
This completely went over my head, will investigate and advise.

**Comment by chrisbenincasa** [comment]
@kagrith any luck?

**Comment by kagrith** [comment]
Kinda!
Sorry, I've been meaning to reply.

I was able to populate my libraries by changing the url slug and keeping
the token the same.

But there seem to be problems when i go to setup channels.
I can add media, but when I try to check on the channel, nothing loads and
it just gives the generic something wrong error, or no error at all and
just a black screen, time codes for the episode won't even load.

I've tried changing the playback type back and forth from plex and direct
off disk.

That seems to have no effect.

Additional, when I try to add flex time, it won't seem to take the save
operation.

I add the flex time, padding, delay, so on, and when I save it claims to
rebuild, but when I try to leave the page it says I have unsaved changes,
no matter how many times I click the save button, or how long I wait.

**Comment by chrisbenincasa** [comment]
Seems like there are 2 problems here (correct me if I'm wrong):

1. You are having a problem with streaming the channel
2. The app is erroneously telling you that you have unsaved changes

For Problem 1, we will need some logs from you server. Presumably something is going with with FFMPEG.

For Problem 2, this is likely a UI issue. Changes are your changes _are_ being saved but the app is wrongly telling you that you have unsaved changes. I can look into this.

**Comment by kagrith** [comment]
I finally found how to get ffmpeg logging and these are the only things that look like a helpful errors.
If you'd like something in particular im happy to oblige.

```
Input #0, concat, from 'http://localhost:8000/playlist?channel=1&audioOnly=false&hls=false':
  Duration: N/A, start: 0.000000, bitrate: 6144 kb/s
  Stream #0:0: Video: rawvideo (Y3[11][10] / 0xA0B3359), yuv420p10le, 1920x1080, SAR 1:1 DAR 16:9, 23.98 tbr, 95904 tbn
      Metadata:
        encoder         : Lavc61.3.100 rawvideo
  Stream #0:1: Audio: pcm_s16le (PSD[16] / 0x10445350), 48000 Hz, 7.1, s16, 6144 kb/s
      Metadata:
        encoder         : Lavc61.3.100 pcm_s16le
[out#0/hls @ 0x324e600] Codec AVOption sc_threshold (Scene change threshold) has not been used for any stream. The most likely reason is either wrong type (e.g. a video option with no video streams) or that it is a private option of some encoder which was not actually used for any stream.
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> hevc (hevc_nvenc))
  Stream #0:1 -> #0:1 (pcm_s16le (native) -> aac (native))
Press [q] to stop, [?] for help
[hls @ 0x324e800] Stream HEVC is not hvc1, you should use tag:v hvc1 to set it.
Output #0, hls, to 'streams/stream_71040716-4fcb-49fe-a105-d5647d8ea353/stream.m3u8':
  Metadata:
    service_provider: "tunarr"
    service_name    : "Channel 1"
    encoder         : Lavf61.1.100
  Stream #0:0: Video: hevc, p010le(progressive), 1920x1080 [SAR 1:1 DAR 16:9], q=2-31, 2000 kb/s, 23.98 fps, 90k tbn
      Metadata:
        encoder         : Lavc61.3.100 hevc_nvenc
      Side data:
        cpb: bitrate max/min/avg: 2000000/0/2000000 buffer size: 2000000 vbv_delay: N/A
  Stream #0:1: Audio: aac, 48000 Hz, stereo, fltp, 192 kb/s
      Metadata:
        encoder         : Lavc61.3.100 aac
[hls @ 0x324e800] Opening 'streams/stream_71040716-4fcb-49fe-a105-d5647d8ea353/data00000.ts' for writing
[hls @ 0x324e800] Opening 'streams/stream_71040716-4fcb-49fe-a105-d5647d8ea353/stream.m3u8.tmp' for writing
[hls @ 0x324e800] Opening 'streams/stream_71040716-4fcb-49fe-a105-d5647d8ea353/master.m3u8' for writing
[hls @ 0x324e800] Opening 'streams/stream_71040716-4fcb-49fe-a105-d5647d8ea353/data00001.ts' for writing
[hls @ 0x324e800] Opening 'streams/stream_71040716-4fcb-49fe-a105-d5647d8ea353/stream.m3u8.tmp' for writing
2024-08-04T21:31:31.262Z [info]: Concat FFMPEG exited due to signal: SIGKILL as expected. {"channel":"71040716-4fcb-49fe-a105-d5647d8ea353"}
```

I've tried playing with the ffmpeg settings, and i can get audio working on the channel, but no video yet, cuda transcoding, on all the codecs, and it just wont chooch.

When i try plugging the tuner into plex i get this error too:
![image](https://github.com/user-attachments/assets/98a5ee6e-5c8b-43f9-9dba-760d3b199bd5)

**Comment by chrisbenincasa** [comment]
Hm, this is the second time I've seen this issue come up. I was just talking with somebody in the discord about this. It seems that for some reason, some builds are not respecting the hls=false parameter when requesting the /playlist endpoint and are starting an HLS session rather than just returning the raw MPEG-TS stream to Plex, which is what Plex expects.

Since you have FFMPEG logging on, are you able to also paste in some more log context. Specifically I'm looking for the log lines that print the raw FFMPEG command that Tunarr is executing.

**Comment by chrisbenincasa** [comment]
I'm looking for the line that begins with "Starting ffmpeg concat process with args"

**Comment by kagrith** [comment]
Another problem is that each time i update the release in unraid, my saved settings get reset.
I'll try and get back to the point where i can get the ffmpeg logging.

**Comment by kagrith** [comment]
Okay, there seems to be some kind of timeout problem with server discovery, but after trying to manually add something, discovery populated repeatedly as i pruned servers i didnt need.

Made a small channel with just one season after the most recent edge-nvidia release since time of writing and after a bit of transcoding, seems to work fine!
Added a second channel with a lot of seasons, and flex padding, and its working too.

My Release
![image](https://github.com/user-attachments/assets/79c51334-796b-480c-b4a2-b54fa515dab4)

Odd error after letting streams cook:
```
[hls @ 0x3270f40] Opening 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/stream.m3u8.tmp' for writing
2024-08-05T12:50:34.581Z [error]: Query error: [object Object] 
    err: {
      "type": "SqliteError",
      "message": "no such column: channel.uuid",
      "stack":
          SqliteError: no such column: channel.uuid
              at Database.prepare (file:///tunarr/server/build/bundle.js:24711:26)
              at SqliteConnection.executeQuery (file:///tunarr/server/build/bundle.js:260912:27)
              at connection.executeQuery (file:///tunarr/server/build/bundle.js:258813:35)
              at file:///tunarr/server/build/bundle.js:253391:40
              at DefaultConnectionProvider.provideConnection (file:///tunarr/server/build/bundle.js:258665:20)
              at async _DefaultQueryExecutor.executeQuery (file:///tunarr/server/build/bundle.js:253390:12)
              at async _SelectQueryBuilderImpl.execute (file:///tunarr/server/build/bundle.js:254830:21)
              at async _SelectQueryBuilderImpl.executeTakeFirst (file:///tunarr/server/build/bundle.js:254834:23)
              at async ChannelDB.getChannelFallbackPrograms (file:///tunarr/server/build/bundle.js:266152:21)
              at async StreamProgramCalculator.createLineupItem (file:///tunarr/server/build/bundle.js:268944:32)
              at async VideoStream.startStream (file:///tunarr/server/build/bundle.js:275063:20)
              at async Object.<anonymous> (file:///tunarr/server/build/bundle.js:275376:31)
      "code": "SQLITE_ERROR"
    }
2024-08-05T12:50:34.582Z [error]: Error on /stream: /stream?channel=2&session=1&audioOnly=false&hls=true&index=1. %O 
    err: {
      "type": "SqliteError",
      "message": "no such column: channel.uuid",
      "stack":
          SqliteError: no such column: channel.uuid
              at Database.prepare (file:///tunarr/server/build/bundle.js:24711:26)
              at SqliteConnection.executeQuery (file:///tunarr/server/build/bundle.js:260912:27)
              at connection.executeQuery (file:///tunarr/server/build/bundle.js:258813:35)
              at file:///tunarr/server/build/bundle.js:253391:40
              at DefaultConnectionProvider.provideConnection (file:///tunarr/server/build/bundle.js:258665:20)
              at async _DefaultQueryExecutor.executeQuery (file:///tunarr/server/build/bundle.js:253390:12)
              at async _SelectQueryBuilderImpl.execute (file:///tunarr/server/build/bundle.js:254830:21)
              at async _SelectQueryBuilderImpl.executeTakeFirst (file:///tunarr/server/build/bundle.js:254834:23)
              at async ChannelDB.getChannelFallbackPrograms (file:///tunarr/server/build/bundle.js:266152:21)
              at async StreamProgramCalculator.createLineupItem (file:///tunarr/server/build/bundle.js:268944:32)
              at async VideoStream.startStream (file:///tunarr/server/build/bundle.js:275063:20)
              at async Object.<anonymous> (file:///tunarr/server/build/bundle.js:275376:31)
      "code": "SQLITE_ERROR"
    }


```

**Comment by kagrith** [comment]
I think at this point i should open up a new issue now that my original one is fixed.

Here is error output of trying to tune into the new channels i just setup:
The first line is what proceeds this error:
![image](https://github.com/user-attachments/assets/52169d7a-e98f-4f20-9fcf-d047384f58e8)
What follows is my attempt to retune in immediately after.
The channel works perfectly from within tunnar, when leveraging the watch channel button.
```
2024-08-05T12:59:45.749Z [info]: Concat FFMPEG exited due to signal: SIGKILL as expected. {"channel":"7e03362a-e1c4-4287-90f7-99e7e365b127"}
2024-08-05T12:59:54.763Z [info]: 
Stream starting. Channel: 1 (LowerDecks)
2024-08-05T12:59:54.792Z [info]: ! Start playback
2024-08-05T12:59:54.792Z [info]: ! Channel: LowerDecks (1)
2024-08-05T12:59:54.792Z [info]: ! Title: Much Ado About Boimler
2024-08-05T12:59:54.792Z [info]: ! From: 0 to: 1502131
2024-08-05T12:59:54.792Z [info]: ! Type: program
{
  disableFillerOverlay: false,
  watermark: {
    url: '',
    enabled: false,
    position: 'bottom-right',
    width: 10,
    verticalMargin: 1,
    horizontalMargin: 1,
    duration: 0,
    fixedSize: false,
    animated: false,
    opacity: 100
  },
  icon: { path: '', width: 0, duration: 0, position: 'bottom-right' },
  name: 'LowerDecks',
  transcoding: undefined,
  number: 1,
  uuid: '7e03362a-e1c4-4287-90f7-99e7e365b127'
} {
  url: '',
  enabled: false,
  position: 'bottom-right',
  width: 10,
  verticalMargin: 1,
  horizontalMargin: 1,
  duration: 0,
  fixedSize: false,
  animated: false,
  opacity: 100
}
[aist#0:1/pcm_s16le @ 0x3267440] Guessed Channel Layout: 5.1=N/A dup=4 drop=0 speed=0.124x    
Input #0, concat, from 'http://localhost:8000/playlist?channel=1&audioOnly=false&hls=false':
  Duration: N/A, start: 0.000000, bitrate: 4608 kb/s
  Stream #0:0: Video: rawvideo (I420 / 0x30323449), yuv420p, 1920x1080, SAR 1:1 DAR 16:9, 23.98 tbr, 48k tbn
      Metadata:
        encoder         : Lavc61.3.100 rawvideo
  Stream #0:1: Audio: pcm_s16le (PSD[16] / 0x10445350), 48000 Hz, 5.1, s16, 4608 kb/s
      Metadata:
        encoder         : Lavc61.3.100 pcm_s16le
[out#0/hls @ 0x3268040] Codec AVOption sc_threshold (Scene change threshold) has not been used for any stream. The most likely reason is either wrong type (e.g. a video option with no video streams) or that it is a private option of some encoder which was not actually used for any stream.
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> h264 (h264_nvenc))
  Stream #0:1 -> #0:1 (pcm_s16le (native) -> aac (native))
Press [q] to stop, [?] for help
Output #0, hls, to 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/stream.m3u8':.388x    
  Metadata:
    service_provider: "tunarr"
    service_name    : "LowerDecks"
    encoder         : Lavf61.1.100
  Stream #0:0: Video: h264, yuv420p(progressive), 1920x1080 [SAR 1:1 DAR 16:9], q=2-31, 2000 kb/s, 23.98 fps, 90k tbn
      Metadata:
        encoder         : Lavc61.3.100 h264_nvenc
      Side data:
        cpb: bitrate max/min/avg: 2000000/0/2000000 buffer size: 2000000 vbv_delay: N/A
  Stream #0:1: Audio: aac, 48000 Hz, stereo, fltp, 192 kb/s
      Metadata:
        encoder         : Lavc61.3.100 aac
2024-08-05T13:00:00.160Z [info]: XMLTV Updated at 8/5/2024, 8:00:00 AM {"task":"update-xmltv"}
[hls @ 0x3268240] Opening 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/data00000.ts' for writing
[hls @ 0x3268240] Opening 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/stream.m3u8.tmp' for writing
[hls @ 0x3268240] Opening 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/master.m3u8' for writing
[hls @ 0x3268240] Opening 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/data00001.ts' for writing
[hls @ 0x3268240] Opening 'streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/stream.m3u8.tmp' for writing
2024-08-05T13:00:19.865Z [info]: Concat FFMPEG exited due to signal: SIGKILL as expected. {"channel":"7e03362a-e1c4-4287-90f7-99e7e365b127"}
frame= 8516 fps=8.3 q=25.0 size=N/A time=00:02:05.58 bitrate=N/A dup=4 drop=0 speed=0.112x    
```

**Comment by kagrith** [comment]
This is a new one:
![image](https://github.com/user-attachments/assets/5cfba446-4116-4ce5-a423-5178eee85a0f)

**Comment by kagrith** [comment]
This error is minutes old at time of writing, doesnt appear to be related to above error, but wanted to include:
`2024-08-05T13:34:54.087Z [error]: CONCAT - FFMPEG ERROR {"code":254,"cmd":"/usr/bin/ffmpeg -hide_banner -threads 1 -fflags +genpts+discardcorrupt+igndts -re -f concat -safe 0 -stream_loop -1 -protocol_whitelist file,http,tcp,https,tcp,tls -probesize 32 -i http://localhost:8000/playlist?channel=2&audioOnly=false&hls=false -flags cgop -sc_threshold 0 -movflags +faststart -map 0:v -map 0:a -c:v h264_nvenc -c:a aac -b:v 2000k -maxrate:v 2000k -bufsize:v 2000k -b:a 192k -maxrate:a 192k -bufsize:a 50k -af aselect=concatdec_select,aresample=async=1 -ac 2 -ar 48k -muxdelay 0 -muxpreload 0 -metadata service_provider=\"tunarr\" -metadata service_name=\"TOS\" -f hls -hls_time 3 -hls_list_size 8 -force_key_frames expr:gte(t,n_forced*3) -hls_delete_threshold 3 -hls_flags independent_segments -hls_flags split_by_time -hls_segment_type mpegts -hls_flags delete_segments -hls_base_url hls/ -hls_segment_filename streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/data%05d.ts -master_pl_name master.m3u8 streams/stream_d78459a2-227f-4e16-a666-6e459e3e4950/stream.m3u8"}`

**Comment by chrisbenincasa** [comment]
Yeah at this point I think it would be helpful to open separate issues for each of these problems; the context can get really muddied reusing the same issue and makes these problems less searchable.

I am aware of the problem from https://github.com/chrisbenincasa/tunarr/issues/602#issuecomment-2269598903 - it has to do with channel fallback program querying and I already have a fix on another branch, just need to isolate it to its own PR.

The playback bug is something that seems to be plaguing others, too. For some reason Tunarr seems to be returning the HLS stream to Plex when it should be returning the MPEG-TS stream. I still haven't gotten a solid repro on this one.

But yes, let's make separate issues for these and we can discuss further there.

**Comment by chrisbenincasa** [comment]
#660 tracking the bad query

**Comment by chrisbenincasa** [comment]
https://github.com/chrisbenincasa/tunarr/issues/662 is tracking the other issue for playback. this was a tricky one that a few folks have called out. I've identified the problem though and am working on a fix

**Comment by chrisbenincasa** [comment]
Both of these should be fixed in 0.5.2
